rails/react-components/src/library/components/permission-forms-v2/students-tab/edit-student-permissions-form.tsx

```diff
@@ -0,0 +1,100 @@
+import React, {useEffect, useState} from "react";
+import { IPermissionForm, IStudent } from "./types";
+import { bulkUpdatePermissionForms } from "./students-table";
+
+import css from "./edit-student-permissions-form.scss";
+
+interface CreateEditPermissionFormProps {
+  student: IStudent;
+  permissionForms: IPermissionForm[];
+  onFormCancel: () => void;
+  onFormSave: () => void;
+  classId: string;
+}
+
+export const EditStudentPermissionsForm = ({ student, permissionForms, onFormCancel, onFormSave, classId }: CreateEditPermissionFormProps) => {
+  const [localPermissions, setLocalPermissions] = useState(student.permission_forms || {});
+  const [idsToRemove, setIdsToRemove] = useState<string[]>([]);
+  const [idsToAdd, setIdsToAdd] = useState<string[]>([]);
+
+  useEffect(() => setLocalPermissions(student.permission_forms), []);
+
+  // whenever a permission checkbox changes we figure out if it represents an add or a remove
+  // we update the form state and the idsToAdd and idsToRemove arrays accordingly
+  const handlePermissionChange = (changedPermissionId: string) => {
+    const shouldAddPermission = !localPermissions.some(lp => lp.id === changedPermissionId);
+    const shouldRemovePermission = localPermissions.some(lp => lp.id === changedPermissionId);
+
+    let newPermissions: IPermissionForm[] = [];
+
+    if (shouldAddPermission) {
+      const permissionToAdd = permissionForms.find(pf => pf.id === changedPermissionId);
+      if (permissionToAdd) {
+        newPermissions = [...localPermissions, permissionToAdd];
+        setIdsToAdd([...idsToAdd, changedPermissionId]);
+      }
+    }
+
+    if (shouldRemovePermission) {
+      newPermissions = localPermissions.filter(lp => lp.id !== changedPermissionId);
+      setIdsToRemove([...idsToRemove, changedPermissionId]);
+    }
+
+    const definedPermissions = newPermissions.filter((p): p is IPermissionForm => p !== undefined);
+    setLocalPermissions(definedPermissions);
+  };
+
+  const handleFormSave = async () => {
+    const response = await bulkUpdatePermissionForms({
+      classId,
+      selectedStudentIds: [student.id.toString()],
+      addFormIds: idsToAdd,
+      removeFormIds: idsToRemove
+    });
```

Comment:
I'd calculate idsToAdd and idsToRemove just here. It's the only moment when we need them.

Also, API could be also called in parent in `onFormSave`, but only if that makes things simpler for some reason. Probably no real difference.